proxy/grpcproxy/lease.go

```diff
@@ -46,42 +95,177 @@ func (lp *leaseProxy) LeaseTimeToLive(ctx context.Context, rr *pb.LeaseTimeToLiv
+		errCh <- lps.run()
 	}()
 
-	for {
-		rr, err := lc.Recv()
-		if err != nil {
-			return err
+	var leaderc <-chan struct{}
+	if md, ok := metadata.FromContext(stream.Context()); ok {
+		v := md[rpctypes.MetadataRequireLeaderKey]
+		if len(v) > 0 && v[0] == rpctypes.MetadataHasLeader {
+			leaderc = lp.lostLeaderNotify()
+		}
+	}
+	select {
+	case <-cctx.Done():
+		return cctx.Err()
+	case <-leaderc:
+		fmt.Printf("leaseProxy leader lost err(%v) \n", rpctypes.ErrNoLeader)
+		return rpctypes.ErrNoLeader
+	case err = <-errCh:
+		return err
+	}
+}
+
+type leaseProxyStream struct {
+	stream pb.Lease_LeaseKeepAliveServer
+	client *clientv3.Client
+
+	wg sync.WaitGroup
+	// mu protects keepAliveLeases
+	mu sync.RWMutex
+	// keepAliveLeases map keeps track of on-going lease renewal IDs and corresponding chans.
+	// the chan tracks number of keepalive requests received for a given lease ID.
+	// it is used to control number of keepalive responses to be sent back to the client
+	keepAliveLeases map[int64]chan struct{}
+
+	ctx    context.Context
+	cancel func()
+}
+
+func (lps *leaseProxyStream) run() error {
+	defer func() {
+		fmt.Printf("leaseProxyStream %v closed \n", lps.stream)
+	}()
+	for lps.ctx.Err() == nil {
+		fmt.Printf("leaseProxyStream %v receiving msg...\n", lps.stream)
+		rr, err := lps.stream.Recv()
+		fmt.Printf("leaseProxyStream %v received msg %v \n", lps.stream, rr)
+		if err == io.EOF {
+			return nil
 		}
-		err = stream.Send(rr)
 		if err != nil {
 			return err
 		}
+		lps.serveKeepAliveRequest(rr)
+	}
+	return lps.ctx.Err()
+}
+
+func (lps *leaseProxyStream) stop() {
+	fmt.Printf("LeaseKeepAlive() stream %v closing... \n", lps.stream)
+	lps.cancel()
+	lps.wg.Wait()
+	fmt.Printf("LeaseKeepAlive() stream %v closed \n", lps.stream)
+}
+
+func (lps *leaseProxyStream) serveKeepAliveRequest(rr *pb.LeaseKeepAliveRequest) {
+	fmt.Printf("leaseProxyStream.serveKeepAliveRequest() req %v ...\n", rr)
+	lps.mu.Lock()
+	defer lps.mu.Unlock()
+	msgsCh, ok := lps.keepAliveLeases[rr.ID]
+	if !ok {
+		msgsCh = make(chan struct{}, 16)
+		lps.keepAliveLeases[rr.ID] = msgsCh
+		lps.wg.Add(1)
+		go func() {
+			defer lps.wg.Done()
+			lps.listeningFromServer(rr.ID)
+		}()
+	}
+	select {
+	case msgsCh <- struct{}{}:
+		fmt.Printf("leaseProxyStream.serveKeepAliveRequest() put items in msgsCh \n")
+	default:
+		// the msgsCh is full probably due to proxy_to_etcd client is down
+		// where proxy is not receiving keepAlive response in which the response is used to drain msgsCh
+		// this is fine because the user client will soon noticed that there aren't any responses
+		// and then user client will close the KeepAlive requests for this lease id.
+	}
+	fmt.Printf("leaseProxyStream.serveKeepAliveRequest() req %v done \n", rr)
+}
+
+func (lps *leaseProxyStream) listeningFromServer(leaseID int64) {
+	fmt.Printf("leaseProxyStream.listeningFromServer() %v ...\n", leaseID)
+	var (
+		msgsCh <-chan struct{}
+		// leaseTTL is a timeout period for this function
```

Comment:
Please don't write essay comments when what's happening isn't very complicated or intricate; it's misleading and makes the code more difficult to read

The code doesn't need to make this assumption anyway, the first select can block indefinitely on a nil ticker channel:
```go
var ticker <-chan time.Time
select {
case <-ticker:
    ...
case ka := <-keepaliveCh:
    ....
    ticker = time.After(ka.TTL)
}
```